ShortURL: Implement /goto with a sub-resource

ryantxu opened the pull request:
This PR adds a custom route `/goto` at the end of a short URL that will do the redirection logic and update the `lastSeenAt` time.

For cloud, we should (eventually) add a rewrite rule in the gateway that sends `/goto/xxx` to the full (not very short!) url

There are still a few odd behaviors with the client -- but I think worth merging now, and then coming back after we [update the SDK ](https://github.com/grafana/grafana/pull/110794)

ryantxu committed:
add goto endpoint

ryantxu committed:
add show path

ryantxu committed:
add show path

ryantxu committed:
Merge remote-tracking branch 'origin/main' into shorturl-goto

ryantxu committed:
merge main

ryantxu committed:
goto

ryantxu committed:
Merge remote-tracking branch 'origin/main' into shorturl-goto

ryantxu committed:
update workspace

ryantxu committed:
merge main

ryantxu committed:
update workspace

ferruvich committed:
Provisioning: add unit and integration tests for finalizer validation (#111012)

* Add unit testS

* add integration tests

Jguer committed:
Authz: Remove legacy API Key permissions (#110860)

* remove API key roles

* remove API key gen

* remove frontend and doc mentions

* restore legacy keygen

* restore codeowners

* prettier

* update swagger

* remove permissions including apikeys

* add migrator for removing deprecated permissions

* add tracing

* update openapi3

* simplify migrator for now

* accesscontrol/migrator: remove batching for deprecated permissions deletion

forsethc committed:
Authz: add logs for monitoring (#110959)

* add logs for monitoring

* add logging around hook enablement

hugohaggmark committed:
Devenv: expose internal faro.receiver server (#111014)

ywzheng1 committed:
RepositoryList: Display managed resource count (#110916)

* RepositoryList: display managed resource count

* display partial managed alert

* Update public/app/features/provisioning/Wizard/hooks/useResourceStats.ts

Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>

* Update public/locales/en-US/grafana.json

Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>

* only count dashboard and folder

---------

Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>

samsch committed:
Accessibility: enable responsive reflow of variables in dashboard edit (#110967)

enable responsive reflow of variables in dashboard edit

iwysiu committed:
Update docs for pdc+sigv4 (#110787)

* Update docs for pdc+sigv4

* Apply suggestions from code review

Co-authored-by: Larissa Wandzura <126723338+lwandz13@users.noreply.github.com>

---------

Co-authored-by: Larissa Wandzura <126723338+lwandz13@users.noreply.github.com>

mgyongyosi committed:
Fix: Fix redirection after login when Grafana is served from subpath (#110889)

Fix short link (/goto) redirection when Grafana is served from subpath

evictorero reviewed (commented):
I've tested it and it works as expected.
We need to decide what to do with the async call and tests before merging it.

harisrozajac committed:
Dashboard Schema V2: Support panel actions (#110842)

* support panel actions

* refactor

* add test; move action transformer to utils

* refactor so v2 headers and queryParams are just a simple record

* update open api

* update actions to be same shape accross all dashboard schemas and add validation on the backend

* cleanup

* update snapshot

* add tests to validation

yuri-tceretian committed:
Make alerting team a sole owner of alerting tests (#111036)

make alerting team sole owner of alerting tests

Commit:
Release: Bump version to 12.3.0-pre (#110974)

* update bump-version

* Add id-token: write

* update generate-token step

* pull-requests -> pull_requests

* clone with token and set right name

* bump version 12.3.0-pre

---------

Co-authored-by: Kevin Minehart <5140827+kminehart@users.noreply.github.com>
Co-authored-by: grafana-delivery-bot[bot] <grafana-delivery-bot[bot]@users.noreply.github.com>

alerting-team[bot] committed:
Alerting: Update alerting module to f2728ab090eed9c6b70057b53239fb370d68e8ed (#111018)

[create-pull-request] automated change

Co-authored-by: santihernandezc <41638679+santihernandezc@users.noreply.github.com>
Co-authored-by: Yuri Tseretyan <yuriy.tseretyan@grafana.com>

leandro-deveikis committed:
CloudMigration: fix flacky test (#111046)

stephaniehingtgen committed:
Provisioning: Fix deletion order (#111043)

owensmallwood committed:
Unified Storage: Adds pruner for kv eventstore (#110785)

* Adds pruner for eventstore - default 24 hours. Adds tests.

* update comment

* remove delay on startup. formatting

* updates log message type and removes useless comment

* caller handles goroutine for runCleanupOldEvents()

* simplify timestamp extraction

* adds config for event pruning interval

* uses start and end key to get all expired events

* remove sort when listing keys in event pruner - order doesnt matter

* use snowflake constants

* log when we delete 0 rows

* pass time.Time to cleanup old events func

fastfrwrd committed:
Table: Fix logic to calculate footer height (#110954)

* Table: Fix logic to calculate footer height

* add non-numeric footer case to gdev

* Update packages/grafana-ui/src/components/Table/TableNG/utils.ts

Co-authored-by: Leon Sorokin <leeoniya@gmail.com>

* Update packages/grafana-ui/src/components/Table/TableNG/TableNG.tsx

Co-authored-by: Leon Sorokin <leeoniya@gmail.com>

---------

Co-authored-by: Leon Sorokin <leeoniya@gmail.com>

fastfrwrd committed:
Table: Restore previous footer behavior of reducers applying to filtered data (#111041)

* Table: Restore previous footer behavior of reducers applying to filtered data

* update e2e to match new behavior

aangelisc committed:
Graphite: Backend metrics expand endpoint (#110678)

* Add lint rules

* Backend decoupling

- Add standalone files
- Add graphite query type
- Add logger to Service
- Create logger in the ProvideService method
- Use a pointer for the HTTP client provider
- Update logger usage everywhere
- Update tracer type
- Replace simplejson with json
- Add dummy CallResource and CheckHealth methods
- Update tests

* Update ConfigEditor imports

* Update types imports

* Update datasource

- Switch to using semver package
- Update imports

* Update store imports

* Update helper imports and notification creation

* Update context import

* Update version numbers and logic

* Copy array_move from core

* Test updates

* Add required files and update plugin.json

* Update core references and packages

* Remove commented code

* Update wire

* Lint

* Fix import

* Copy null type

* More lint

* Update snapshot

* Refactor backend

- Split query logic into separate file
- Move utils to separate file

* Add health-check logic

- Support backend healthcheck if the FF is enabled

* Remove query import support as unneeded

* Add test

* Add util function for decoding responses

* Add events types

* Add resource handler

* Add events handler and generic resource req handler

* Tests

* Update frontend

- Add types
- Update events function to support backend requests

* Lint and typing

* Lint

* Add metrics find endpoint

- Add types
- Add generic response parser
- Add endpoint
- Tests

* Update FE functoin to use backend endpoint

* Lint

* Simplify request

* Update test

* Metrics expand type

* Extract shared logic and add metric expand endpoint

* Update tests

* Call metric expand from backend

* Add tests

* Review

* Review

* Fix packages

* Format

* Fix merge issues

* Review

* Fix undefined values

* Extract request creation

- Add method for create requests generically with tests
- Replace usage in query method
- Update usages in resource handlers
- Update tests
- Update types

moustafab committed:
Alerting: Rename expression elements of Rules APIs (#110914)

This renames `data` to `expressions` for clarity in the rules apis.

Also makes certain fields that are redundant optional in the case of pure expressions, so that users don't have to specify them when they are not needed (e.g. not datasource queries).

aangelisc committed:
Graphite: Backend functions endpoint (#110771)

* Add lint rules

* Backend decoupling

- Add standalone files
- Add graphite query type
- Add logger to Service
- Create logger in the ProvideService method
- Use a pointer for the HTTP client provider
- Update logger usage everywhere
- Update tracer type
- Replace simplejson with json
- Add dummy CallResource and CheckHealth methods
- Update tests

* Update ConfigEditor imports

* Update types imports

* Update datasource

- Switch to using semver package
- Update imports

* Update store imports

* Update helper imports and notification creation

* Update context import

* Update version numbers and logic

* Copy array_move from core

* Test updates

* Add required files and update plugin.json

* Update core references and packages

* Remove commented code

* Update wire

* Lint

* Fix import

* Copy null type

* More lint

* Update snapshot

* Refactor backend

- Split query logic into separate file
- Move utils to separate file

* Add health-check logic

- Support backend healthcheck if the FF is enabled

* Remove query import support as unneeded

* Add test

* Add util function for decoding responses

* Add events types

* Add resource handler

* Add events handler and generic resource req handler

* Tests

* Update frontend

- Add types
- Update events function to support backend requests

* Lint and typing

* Lint

* Add metrics find endpoint

- Add types
- Add generic response parser
- Add endpoint
- Tests

* Update FE functoin to use backend endpoint

* Lint

* Simplify request

* Update test

* Metrics expand type

* Extract shared logic and add metric expand endpoint

* Update tests

* Call metric expand from backend

* Rename type for clarity

* Add get resource req handler

* Refactor doGraphiteRequest, parseResponse

Update tests

* Migrate functions endpoint to backend

* Add tests

* Review

* Review

* Fix packages

* Format

* Fix merge issues

* Review

* Fix undefined values

* Extract request creation

- Add method for create requests generically with tests
- Replace usage in query method
- Update usages in resource handlers
- Update tests
- Update types

* Lint

* Lint

aangelisc committed:
Graphite: Backend tags autocomplete endpoint (#110772)

* Add lint rules

* Backend decoupling

- Add standalone files
- Add graphite query type
- Add logger to Service
- Create logger in the ProvideService method
- Use a pointer for the HTTP client provider
- Update logger usage everywhere
- Update tracer type
- Replace simplejson with json
- Add dummy CallResource and CheckHealth methods
- Update tests

* Update ConfigEditor imports

* Update types imports

* Update datasource

- Switch to using semver package
- Update imports

* Update store imports

* Update helper imports and notification creation

* Update context import

* Update version numbers and logic

* Copy array_move from core

* Test updates

* Add required files and update plugin.json

* Update core references and packages

* Remove commented code

* Update wire

* Lint

* Fix import

* Copy null type

* More lint

* Update snapshot

* Refactor backend

- Split query logic into separate file
- Move utils to separate file

* Add health-check logic

- Support backend healthcheck if the FF is enabled

* Remove query import support as unneeded

* Add test

* Add util function for decoding responses

* Add events types

* Add resource handler

* Add events handler and generic resource req handler

* Tests

* Update frontend

- Add types
- Update events function to support backend requests

* Lint and typing

* Lint

* Add metrics find endpoint

- Add types
- Add generic response parser
- Add endpoint
- Tests

* Update FE functoin to use backend endpoint

* Lint

* Simplify request

* Update test

* Metrics expand type

* Extract shared logic and add metric expand endpoint

* Update tests

* Call metric expand from backend

* Rename type for clarity

* Add get resource req handler

* Refactor doGraphiteRequest, parseResponse

Update tests

* Migrate functions endpoint to backend

* Support tags autocomplete in backend

- Add tests
- Add types
- Remove unneeded comments

* Add tests

* Review

* Review

* Fix packages

* Format

* Fix merge issues

* Review

* Fix undefined values

* Extract request creation

- Add method for create requests generically with tests
- Replace usage in query method
- Update usages in resource handlers
- Update tests
- Update types

evictorero committed:
ShortURL: Use the new k8s api in the frontend (#110537)

grafana-pr-automation[bot] committed:
I18n: Download translations from Crowdin (#111052)

New Crowdin translations by GitHub Action

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ryantxu committed:
make sync

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=110972) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=110972) before we can accept your contribution.<br/>**17** out of **19** committers have signed the CLA.<br/><br/>:white_check_mark: ferruvich<br/>:white_check_mark: ryantxu<br/>:white_check_mark: forsethc<br/>:white_check_mark: hugohaggmark<br/>:white_check_mark: ywzheng1<br/>:white_check_mark: iwysiu<br/>:white_check_mark: samsch<br/>:white_check_mark: leandro-deveikis<br/>:white_check_mark: harisrozajac<br/>:white_check_mark: mgyongyosi<br/>:white_check_mark: Jguer<br/>:white_check_mark: fastfrwrd<br/>:white_check_mark: aangelisc<br/>:white_check_mark: owensmallwood<br/>:white_check_mark: evictorero<br/>:white_check_mark: moustafab<br/>:white_check_mark: stephaniehingtgen<br/>:x: alerting-team[bot]<br/>:x: grafana-pr-automation[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=110972) it.</sub>

ryantxu committed:
Merge remote-tracking branch 'origin/main' into shorturl-goto

ryantxu committed:
update tests

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=110972) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=110972) before we can accept your contribution.<br/>**17** out of **19** committers have signed the CLA.<br/><br/>:white_check_mark: ryantxu<br/>:white_check_mark: ferruvich<br/>:white_check_mark: forsethc<br/>:white_check_mark: Jguer<br/>:white_check_mark: hugohaggmark<br/>:white_check_mark: samsch<br/>:white_check_mark: iwysiu<br/>:white_check_mark: harisrozajac<br/>:white_check_mark: mgyongyosi<br/>:white_check_mark: ywzheng1<br/>:white_check_mark: evictorero<br/>:white_check_mark: stephaniehingtgen<br/>:white_check_mark: aangelisc<br/>:white_check_mark: fastfrwrd<br/>:white_check_mark: owensmallwood<br/>:white_check_mark: moustafab<br/>:white_check_mark: leandro-deveikis<br/>:x: alerting-team[bot]<br/>:x: grafana-pr-automation[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=110972) it.</sub>

ryantxu committed:
goto client

ryantxu committed:
merge main

ryantxu commented:
@evictorero -- i think this is good to merge.  I disabled the mode 3+4 tests -- we can stick them back once we have a client with UpdateStatus

evictorero reviewed (approved):
LGTM! Tested and work as expected. Created this [task](https://github.com/grafana/grafana/issues/111095) to keep track of the remaining work.

IfSentient commented on apps/shorturl/pkg/app/app.go:
```diff
@@ -61,6 +71,48 @@ func New(cfg app.Config) (app.App, error) {
+						Method: "GET",
+						Path:   "goto",
+					}: func(ctx context.Context, w app.CustomRouteResponseWriter, req *app.CustomRouteRequest) error {
+						url, _, found := strings.Cut(req.URL.Path, "/apis/") // This will be settings.AppURL
```
Sorry, missed this earlier! We [pass the URL through from the one given to us in the request](https://github.com/grafana/grafana-app-sdk/blob/main/k8s/apiserver/connector.go#L163), so something in the kubernetes machinery may be messing with it, I'll have to poke around.
